cpp/include/raft/spectral/kmeans.hpp

```diff
@@ -0,0 +1,898 @@
+  for (i = 1; i < k; ++i) {
+    // Choose ith centroid
+    if (chooseNewCentroid(handle, thrust_exec_policy, n, d, k, uniformDist(rng),
+                          obs, dists, centroids + IDX(0, i, d)))
+      WARNING("error in k-means++ (could not pick centroid)");
+
+    // Compute distances from ith centroid
+    CUDA_TRY(cudaMemsetAsync(dists + n, 0, n * sizeof(ValueType_), stream));
+    computeDistances<<<gridDim_warp, blockDim_warp, 0, stream>>>(
+      n, d, 1, obs, centroids + IDX(0, i, d), dists + n);
+    CHECK_CUDA(stream);
+
+    // Recompute minimum distances
+    minDistances2<<<gridDim_block, BLOCK_SIZE, 0, stream>>>(n, dists, dists + n,
+                                                            codes, i);
+    CHECK_CUDA(stream);
+  }
+
+  // Compute cluster sizes
+  CUDA_TRY(cudaMemsetAsync(clusterSizes, 0, k * sizeof(IndexType_), stream));
+  computeClusterSizes<<<gridDim_block, BLOCK_SIZE, 0, stream>>>(n, k, codes,
+                                                                clusterSizes);
+  CHECK_CUDA(stream);
+
+  return 0;
+}
+
+/// Find cluster centroids closest to observation vectors
+/** Distance is measured with Euclidean norm.
+ *
+ *  @param n Number of observation vectors.
+ *  @param d Dimension of observation vectors.
+ *  @param k Number of clusters.
+ *  @param obs (Input, device memory, d*n entries) Observation
+ *    matrix. Matrix is stored column-major and each column is an
+ *    observation vector. Matrix dimensions are d x n.
+ *  @param centroids (Input, device memory, d*k entries) Centroid
+ *    matrix. Matrix is stored column-major and each column is a
+ *    centroid. Matrix dimensions are d x k.
+ *  @param dists (Output, device memory, n*k entries) Workspace. On
+ *    exit, the first n entries give the square of the Euclidean
+ *    distance between observation vectors and the closest centroid.
+ *  @param codes (Output, device memory, n entries) Cluster
+ *    assignments.
+ *  @param clusterSizes (Output, device memory, k entries) Number of
+ *    points in each cluster.
+ *  @param residual_host (Output, host memory, 1 entry) Residual sum
+ *    of squares of assignment.
+ *  @return Zero if successful. Otherwise non-zero.
+ */
+template <typename IndexType_, typename ValueType_, typename ThrustExePolicy>
+static int assignCentroids(
+  handle_t const& handle, ThrustExePolicy thrust_exec_policy, IndexType_ n,
+  IndexType_ d, IndexType_ k, const ValueType_* __restrict__ obs,
+  const ValueType_* __restrict__ centroids, ValueType_* __restrict__ dists,
+  IndexType_* __restrict__ codes, IndexType_* __restrict__ clusterSizes,
+  ValueType_* residual_host) {
+  // CUDA grid dimensions
+  dim3 blockDim, gridDim;
+
+  auto cublas_h = handle.get_cublas_handle();
+  auto stream = handle.get_stream();
+
+  // Compute distance between centroids and observation vectors
+  CUDA_TRY(cudaMemsetAsync(dists, 0, n * k * sizeof(ValueType_), stream));
+  blockDim.x = WARP_SIZE;
```

Comment:
I'm not sure what is meant here by "consolidate". Could you please elaborate?

Reply:
Rather than redefining the same dim3  everywhere for grid and block sizes, my suggestion is have a utility function or just a constant somewhere.

Reply:
Still not clear, I'm afraid. It's not the same `dim3`, the values are slightly different. But even if they where, are you suggesting defining a global `dim3` just to avoid repetition? Polluting the namespace with global variables like that is not good practice. Or, maybe I'm misunderstanding.

Could you please post an example of what is desired here? Thanks in advance.